proxy.go

```diff
@@ -0,0 +1,284 @@
+package tunnel
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/koding/logging"
+	"github.com/koding/tunnel/proto"
+)
+
+// ProxyFunc is responsible for forwarding a remote connection to local server and writing the response back.
+type ProxyFunc func(remote net.Conn, msg *proto.ControlMessage)
+
+var (
+	// DefaultProxyFuncs holds global default proxy functions for all transport protocols.
+	DefaultProxyFuncs = ProxyFuncs{
+		HTTP: new(HTTPProxy).Proxy,
+		TCP:  new(TCPProxy).Proxy,
+		WS:   new(HTTPProxy).Proxy,
+	}
+	// DefaultProxy is a ProxyFunc that uses DefaultProxyFuncs.
+	DefaultProxy = Proxy(ProxyFuncs{})
+)
+
+// ProxyFuncs is a collection of ProxyFunc.
+type ProxyFuncs struct {
+	// HTTP is custom implementation of HTTP proxing.
+	HTTP ProxyFunc
+	// TCP is custom implementation of TCP proxing.
+	TCP ProxyFunc
+	// WS is custom implementation of web socket proxing.
+	WS ProxyFunc
+}
+
+// Proxy returns a ProxyFunc that uses custom function if provided, otherwise falls back to DefaultProxyFuncs.
+func Proxy(p ProxyFuncs) ProxyFunc {
+	return func(remote net.Conn, msg *proto.ControlMessage) {
+		var f ProxyFunc
+		switch msg.Protocol {
+		case proto.HTTP:
+			f = DefaultProxyFuncs.HTTP
+			if p.HTTP != nil {
+				f = p.HTTP
+			}
+		case proto.TCP:
+			f = DefaultProxyFuncs.TCP
+			if p.TCP != nil {
+				f = p.TCP
+			}
+		case proto.WS:
+			f = DefaultProxyFuncs.WS
+			if p.WS != nil {
+				f = p.WS
+			}
+		}
+
+		if f == nil {
+			logging.Error("Could not determine proxy function for %v", msg)
+			remote.Close()
+		}
+
+		f(remote, msg)
+	}
+}
+
+////////////////
+// HTTP proxy //
```

Comment:
If you needed to partition single file with such headers, it may mean we'd need to split it into several smaller files.